protobuf-provider/src/main/java/io/confluent/kafka/schemaregistry/protobuf/ProtobufSchema.java

```diff
@@ -0,0 +1,707 @@
+      }
+    }
+    for (ReservedElement reserved : messageElem.getReserveds()) {
+      for (Object elem : reserved.getValues()) {
+        if (elem instanceof String) {
+          message.addReservedName((String) elem);
+        } else if (elem instanceof Integer) {
+          int tag = (Integer) elem;
+          message.addReservedRange(tag, tag);
+        } else if (elem instanceof Range) {
+          Range<Integer> range = (Range<Integer>) elem;
+          message.addReservedRange(range.lowerEndpoint(), range.upperEndpoint());
+        } else {
+          throw new IllegalStateException("Unsupported reserved type: " + elem.getClass()
+              .getName());
+        }
+      }
+    }
+    return message.build();
+  }
+
+  private static Optional<OptionElement> findOption(String name, List<OptionElement> options) {
+    return options.stream().filter(o -> o.getName().equals(name)).findFirst();
+  }
+
+  private static EnumDefinition toDynamicEnum(EnumElement enumElem) {
+    Boolean allowAlias = findOption("allow_alias", enumElem.getOptions()).map(o -> Boolean.valueOf(o
+        .getValue()
+        .toString())).orElse(null);
+    EnumDefinition.Builder enumer = EnumDefinition.newBuilder(enumElem.getName(), allowAlias);
+    for (EnumConstantElement constant : enumElem.getConstants()) {
+      enumer.addValue(constant.getName(), constant.getTag());
+    }
+    return enumer.build();
+  }
+
+  public ProtoFileElement rawSchema() {
+    return schemaObj;
+  }
+
+  @Override
+  public String schemaType() {
+    return TYPE;
+  }
+
+  @Override
+  public String name() {
+    return firstMessage().getName();
+  }
+
+  @Override
+  public String canonicalString() {
+    // Remove full-line comments, such as the location
+    return schemaObj.toSchema().replaceAll("^//.*?\\n", "");
+  }
+
+  public Integer version() {
+    return version;
+  }
+
+  @Override
+  public List<SchemaReference> references() {
+    return references;
+  }
+
+  public Map<String, String> resolvedReferences() {
+    return dependencies.entrySet()
+        .stream()
+        .collect(Collectors.toMap(Map.Entry::getKey, e -> e.getValue().toSchema()));
+  }
+
+  public Map<String, ProtoFileElement> dependencies() {
+    return dependencies;
+  }
+
+  @Override
+  public boolean isBackwardCompatible(ParsedSchema previousSchema) {
+    if (!schemaType().equals(previousSchema.schemaType())) {
+      return false;
+    }
+    final List<Difference> differences = SchemaDiff.compare(
+        ((ProtobufSchema) previousSchema).schemaObj,
+        schemaObj
+    );
+    final List<Difference> incompatibleDiffs = differences.stream()
+        .filter(diff -> !SchemaDiff.COMPATIBLE_CHANGES.contains(diff.getType()))
+        .collect(Collectors.toList());
+    boolean isCompatible = incompatibleDiffs.isEmpty();
+    if (!isCompatible) {
+      log.warn("Found incompatible change: {}", incompatibleDiffs.get(0));
+    }
+    return isCompatible;
```

Comment:
Should we log all the incompatible changes ? If not, can we stop at the first one, instead of computing all of them ?

Reply:
I'll log the remaining at debug level.